playbook/actions/gitops.go

```diff
@@ -0,0 +1,236 @@
+	"github.com/flanksource/gomplate/v3"
+	v1 "github.com/flanksource/incident-commander/api/v1"
+	"github.com/flanksource/incident-commander/pkg/clients/git"
+	"github.com/flanksource/incident-commander/pkg/clients/git/connectors"
+	gitv5 "github.com/go-git/go-git/v5"
+)
+
+type GitOps struct {
+	workTree *gitv5.Worktree
+	spec     *connectors.GitopsAPISpec
+	env      TemplateEnv
+
+	shouldCreatePR bool
+}
+
+type GitOpsActionResult struct {
+	CreatedPR int `json:"createdPR,omitempty"`
+}
+
+func (t *GitOps) Run(ctx context.Context, action v1.GitOpsAction, env TemplateEnv) (*GitOpsActionResult, error) {
+	var response GitOpsActionResult
+
+	if len(action.Patches) == 0 && len(action.Files) == 0 {
+		logger.Warnf("no patches or files specified on gitops action")
+		return nil, nil
+	}
+
+	t.env = env
+
+	if err := t.generateSpec(ctx, action); err != nil {
+		return nil, err
+	}
+
+	connector, workTree, err := t.cloneRepo(ctx, action)
+	if err != nil {
+		return nil, fmt.Errorf("failed to clone repo: %w", err)
+	}
+	t.workTree = workTree
+
+	if err := t.applyPatches(ctx, action); err != nil {
+		return nil, fmt.Errorf("failed to apply patches: %w", err)
+	}
+
+	if err := t.modifyFiles(ctx, action); err != nil {
+		return nil, fmt.Errorf("failed to modify files: %w", err)
+	}
+
+	if err := git.CommitAndPush(ctx, connector, workTree, t.spec); err != nil {
+		return nil, fmt.Errorf("failed to commit and push: %w", err)
+	}
+
+	if t.shouldCreatePR {
+		prNumber, err := t.createPR(ctx, connector, workTree)
+		if err != nil {
+			return nil, err
+		}
+		response.CreatedPR = prNumber
+	}
+
+	return &response, nil
+}
+
+// generateSpec generates the spec for the git client from the action
+func (t *GitOps) generateSpec(ctx context.Context, action v1.GitOpsAction) error {
+	var err error
+
+	if action.Repo.Base == "" {
+		action.Repo.Base = "master"
+	}
+
+	if action.Repo.Branch == "" {
+		action.Repo.Branch = action.Repo.Base
+	} else {
+		action.Repo.Branch, err = gomplate.RunTemplate(t.env.AsMap(), gomplate.Template{Template: action.Repo.Branch})
+		if err != nil {
+			return err
+		}
+	}
+
+	t.spec = &connectors.GitopsAPISpec{
+		Repository: action.Repo.URL,
+		Base:       action.Repo.Base,
+		Branch:     action.Repo.Branch,
+		CommitMsg:  action.Commit.Message,
+		User:       action.Commit.AuthorName,
+		Email:      action.Commit.AuthorEmail,
+	}
+
+	if t.spec.CommitMsg != "" {
+		t.spec.CommitMsg, err = gomplate.RunTemplate(t.env.AsMap(), gomplate.Template{Template: t.spec.CommitMsg})
+		if err != nil {
+			return err
+		}
+	}
+
+	if action.Repo.Connection != "" {
+		conn, err := ctx.HydrateConnectionByURL(action.Repo.Connection)
+		if err != nil {
+			return err
+		}
+
+		switch conn.Type {
+		case models.ConnectionTypeGithub:
+			t.spec.GITHUB_TOKEN = conn.Password
+			t.shouldCreatePR = true
+
+		case models.ConnectionTypeAzureDevops:
+			// TODO: Azure devops connection doesn't have git credentials ...?
+
+		case models.ConnectionTypeGit:
```

Comment:
the would be case git or http

Bot http and  git connection can specify a username and password, and git can specify a SSH clone key (certificate)